Assignments/Program.cs

```diff
@@ -1 +1,70 @@
-﻿Console.WriteLine("Hello, World!");
+﻿using System;
+using System.Numerics;
+using static System.ConsoleColor;
+
+namespace Spark;
+
+public class T3 {
+   public static void Main () {
+      Console.WriteLine ("\x1B[4m" + "Fibonacci Series Printer:-" + "\x1B[0m");
+      List<int> testCases = new () { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 15, 20, 25};
+      foreach (int i in testCases) {
+         Console.Write ($"\nf{i} ==> ");
+         Console.ForegroundColor = Yellow;
+         FiboSeries (i);
+         Console.ResetColor ();
+      }
+      Console.Write ("\n\nTry Yourself! Enter the length of series, l = ");
+      if (int.TryParse (Console.ReadLine (), out int l)) {
+         Console.Write ($"\nf{l} ==> ");
+         Console.ForegroundColor = Yellow;
+         FiboSeries (l);
+         Console.ResetColor ();
+      }
+   }
+
+   /// <summary>Prints the series of n Fibonacci numbers</summary>
+   public static void FiboSeries (int count) {
+      if (count <= 0) Console.WriteLine ("Please enter a +ve value");
+      BigInteger f0 = 0, f1 = 1;
+      while (count > 0) {
+         Console.Write ($"{f0}  ");
+         (f0, f1) = (f1, f0 + f1);
+         count--;
+      }
+   }
+
+   #region Nth Fibonacci Number:
+   /// <summary>Returns nth Fibonacci number using caching</summary>
+   public static BigInteger NthFib3 (int n) {
+      Console.ForegroundColor = Green;
+      BigInteger f0 = 0, f1 = 1;
+      while (0 <= n--) {
+         if (n == 0) return f0;
+         (f0, f1) = (f1, f0 + f1);
+      }
+      return f0;
+   }
+
+   /// <summary>Returns nth Fibonacci number using golden ratio</summary>
+   public static BigInteger NthFib1 (int n) {
+      //This method works upto (n <= 71)
+      double sqrt5 = Math.Sqrt (5);
+      double phi = (sqrt5 + 1) / 2; n--;
+      return (BigInteger)(Math.Ceiling (Math.Pow (phi, n) - Math.Pow (1 - phi, n)) / sqrt5);

```

Comment:
This won't work always. See [here](https://stackoverflow.com/questions/18016879/error-in-converting-double-to-biginteger).

Reply:
Okay. Anyway, this method works fine till 71th fib number. (Already mentioned in the comment).

Reply:
Then you could have used ulong as ulong.Max is greater than 71th fib number.

Reply:
Here the limit at 71st fib num is due to insufficient number of decimal accuracies in the double phi and sqrt5.

Reply:
@vikrampg1710-tmm Sorry, I don't understand what you are trying to say. Are you saying if I use ulong, the result won't be correct?

Reply:
I am just saying you could replace BigInteger with ulong. Not double with ulong.

Reply:
Yes!

Reply:
> Okay. Anyway, this method works fine till 71st fib number. (Already mentioned in the comment).

Sorry, little correction: This method works fine till n = 72.

Reply:
@vikrampg1710-tmm I am not sure what you are trying to achieve. I said double will lose precision if you cast it to BigInetger. Take away is that you should not cast double to BigInteger as mentioned in StackOverflow site. I just informed you so that in future if you try to use it you would be aware. But you are arguing other than that. Please understand that I am just trying to help you and not pointing out your mistake. You have to take reviews constructively.

Reply:
> Yes! To compute 73rd fib number using psi, we need more accurate decimal precision than double-precision float. Here the precision of psi and sqrt5 is not sufficient. Methods: ![image](https://user-images.githubusercontent.com/117505871/275249391-3c0af234-86d3-437a-a3e8-6ce9480974eb.png)
> 
> Output: ![image](https://user-images.githubusercontent.com/117505871/275249182-878cca9f-d1b6-4d80-9493-2dec79072bf3.png)

Are you saying that if you use BigInteger this problem won't occur.

Reply:
No, this problem will occur irrespective of return type, ulong or BigInteger.
![image](https://github.com/vikrampg1710-tmm/Training/assets/117505871/5801c05a-0c21-4fba-ad7f-b384b5ac98b4)


> @vikrampg1710-tmm I am not sure what you are trying to achieve. I said double will lose precision if you cast it to BigInetger. Take away is that you should not cast double to BigInteger as mentioned in StackOverflow site. I just informed you so that in future if you try to use it you would be aware. But you are arguing other than that. Please understand that I am just trying to help you and not pointing out your mistake. You have to take reviews constructively.

Sorry anna. I didn't intent to argue or prove something.  I thought you misunderstood my code, so I tried to give valid reasons.